File: 2020/day06/solution.py
import sys

def reduce(function, iterable, initializer=None):
  it = iter(iterable)
  if initializer is None:
    value = next(it)
  else:
    value = initializer
  for element in it:
    value = function(value, element)
  return value

with open(sys.argv[1]) as f:
    r = f.read().split("\n\n")
    print("🎄 => " + str(sum(map(lambda z: len(set(z.replace("\n",""))), r))))
    print("🎄 => " + str(sum(map(lambda z: len(reduce(lambda x, y: x & y, map(set, z.splitlines()))), r))))

File: 2020/day06/solution.rb
require "set"
INFILE = File.read('input').split("\n\n")
puts "🎄 => " + "#{ INFILE.map { |group| group.gsub("\n", '').split('').to_set.length }.reduce(:+) }"
puts "🎄 => " + "#{ INFILE.map { |group| group.split("\n").map { |list| list.split('') }.reduce(:&).length }.reduce(:+) }"
